race-buffer/src/writer.rs

```diff
@@ -0,0 +1,501 @@
+            if unsafe { self.read_storage(self.write_seqn).is_prefix() } {
+                // Reading storage is safe because buffer has already looped once
+                overwritten = unsafe {
+                    OverwrittenEntry::Double(
+                        self.read_storage(self.overwrite_seqn),
+                        self.read_storage(seqn_add(self.overwrite_seqn, 1, self.seqn_mod)),
+                    )
+                };
+                // If overwriting double, increment overwrite sequence number by 2
+                self.overwrite_seqn = seqn_add(self.overwrite_seqn, 2, self.seqn_mod);
+                // Prevent writes from getting reordered
+                fence(Ordering::Release);
+            } else {
+                // Reading storage is safe because buffer has already looped once
+                overwritten =
+                    unsafe { OverwrittenEntry::Single(self.read_storage(self.overwrite_seqn)) };
+                self.overwrite_seqn = seqn_add(self.overwrite_seqn, 1, self.seqn_mod);
+                // Prevent writes from getting reordered
+                fence(Ordering::Release);
+            }
+        } else {
+            overwritten = OverwrittenEntry::None;
+        }
+        self.write_to_storage(self.write_seqn, entry);
+        // Prevent writes from getting reordered
+        fence(Ordering::Release);
+        self.write_seqn = seqn_add(self.write_seqn, 1, self.seqn_mod);
+        // Prevent writes from getting reordered
+        fence(Ordering::Release);
+        overwritten
+    }
+
+    /// Write double entry in single borrow, returning overwritten entry
+    pub fn write_double(
+        &mut self,
+        prefix: E,
+        suffix: E,
+    ) -> (OverwrittenEntry<E>, OverwrittenEntry<E>) {
+        let first_overwritten = self.write(prefix);
+        let second_overwritten = self.write(suffix);
+        (first_overwritten, second_overwritten)
+    }
+
+    /// Read entry at given sequence number. Entry may or may not currently be present in the buffer.
+    #[inline]
+    pub fn read(&self, mut read_seqn: u32) -> PossiblyMissed<E> {
+        if read_seqn >= self.get_seqn_mod() {
+            read_seqn %= self.get_seqn_mod();
+        }
+        let num_missed = num_missed(
+            read_seqn,
+            self.overwrite_seqn,
+            self.write_seqn,
+            self.seqn_mod,
+        );
+        // If num_missed is greater than zero, entry is not present. If the read sequence number is caught
+        // up to the write seqn, then the next entry has not been written yet
+        if num_missed == 0 && read_seqn != self.write_seqn {
+            PossiblyMissed::Entry(unsafe { self.read_storage(read_seqn) })
+        } else {
+            PossiblyMissed::Missed
+        }
+    }
+
+    /// Iterate over the entries, where None represents a missed entry
+    #[inline]
+    pub fn iter<'b>(&'b self, start_seqn: u32) -> RaceBufferIter<'a, 'b, E> {
```

Comment:
It feels strange to have to supply a starting index to an iterator, when for the probe use case we always need to start at the "start."  

I get that for fully separated reader/writer pairs we track the cursors distinctly and this API makes sense in that context, but perhaps there's a slightly higher level abstraction we could provide around RaceBuffer that manages a read cursor to give simplified access to reading from the head, deleting from the head, and writing to the tail.